File: src/core/playlist.ts
import { fetchWithRetry } from "./http.js";
import { parseTimestamp, stepTimestamp } from "./time.js";

/**
 * Source map in this file:
 * - playlistCreateUrl: rajiko/modules/timeshift.js (playlist_create_url)
 * - buildSegmentUrls: rajiko/modules/timeshift.js (downloadtimeShift seek/chunklist loop)
 */
/**
 * Source: rajiko/modules/timeshift.js -> playlist_create_url
 */
export async function playlistCreateUrl(stationId: string): Promise<string> {
  const resp = await fetchWithRetry(
    `https://radiko.jp/v3/station/stream/pc_html5/${stationId}.xml`,
    undefined,
    {
      retries: 3,
      delayMs: 300,
    },
  );
  if (!resp.ok) {
    throw new Error(`station stream xml failed: ${resp.status}`);
  }
  const xml = await resp.text();
  const regex =
    /((areafree="0".*?timefree="1")|(timefree="1".*?areafree="0")).*\n.*<playlist_create_url>(.*?)<\/playlist_create_url>/gm;
  const match = regex.exec(xml);
  if (match?.[4]) {
    return match[4];
  }
  return "https://tf-f-rpaa-radiko.smartstream.ne.jp/tf/playlist.m3u8";
}

/**
 * Source: rajiko/modules/timeshift.js -> downloadtimeShift (seek/chunklist logic)
 */
export async function buildSegmentUrls(input: {
  stationId: string;
  ft: string;
  to: string;
  token: string;
  areaId: string;
}): Promise<string[]> {
  const { stationId, ft, to, token, areaId } = input;
  const FIXED_SEEK = 300;
  const base = new URL(await playlistCreateUrl(stationId));
  base.searchParams.set("lsid", randomHex(32));
  base.searchParams.set("station_id", stationId);
  base.searchParams.set("l", String(FIXED_SEEK));
  base.searchParams.set("start_at", ft);
  base.searchParams.set("end_at", to);
  base.searchParams.set("type", "b");
  base.searchParams.set("ft", ft);
  base.searchParams.set("to", to);

  const links: string[] = [];
  let seek = ft;
  const endDate = parseTimestamp(to);
  while (parseTimestamp(seek) < endDate) {
    base.searchParams.set("seek", seek);
    const playlistResp = await fetchWithRetry(
      base.toString(),
      {
        headers: {
          "X-Radiko-AreaId": areaId,
          "X-Radiko-AuthToken": token,
        },
      },
      { retries: 3, delayMs: 250 },
    );
    const playlistText = await playlistResp.text();
    if (
      !playlistResp.ok ||
      playlistResp.status === 403 ||
      playlistText === "expired"
    ) {
      throw new Error(
        `playlist request failed at seek=${seek}: ${playlistResp.status}`,
      );
    }

    const detailUrl = firstDataLine(playlistText);
    const chunkResp = await fetchWithRetry(detailUrl, undefined, {
      retries: 3,
      delayMs: 250,
    });
    if (!chunkResp.ok) {
      throw new Error(`chunklist request failed: ${chunkResp.status}`);
    }
    const chunkText = await chunkResp.text();
    links.push(...allDataLines(chunkText));
    seek = stepTimestamp(seek, FIXED_SEEK);
  }

  return links;
}

function firstDataLine(m3u8: string): string {
  const line = allDataLines(m3u8)[0];
  if (!line) {
    throw new Error("m3u8 has no media lines");
  }
  return line;
}

function allDataLines(m3u8: string): string[] {
  return m3u8
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line !== "" && !line.startsWith("#"));
}

function randomHex(length: number): string {
  const hex = "0123456789abcdef";
  let out = "";
  for (let i = 0; i < length; i += 1) {
    out += hex[Math.floor(Math.random() * hex.length)];
  }
  return out;
}

File: internal/domain/playlist.go
package domain

import (
	"context"
	"fmt"
	"regexp"
	"strings"

	"rajidou/internal/netx"
	"rajidou/internal/util"
)

// Source map in this file:
// - playlistCreateURL behavior: rajiko/modules/timeshift.js (playlist_create_url)
// - seek/chunklist loop: rajiko/modules/timeshift.js (downloadtimeShift)
// PlaylistBuilder expands timeshift metadata into ordered AAC segment URLs.
type PlaylistBuilder struct {
	net *netx.Client
}

// NewPlaylistBuilder creates a playlist builder backed by the shared HTTP client.
func NewPlaylistBuilder(net *netx.Client) *PlaylistBuilder {
	return &PlaylistBuilder{net: net}
}

func (p *PlaylistBuilder) playlistCreateURL(ctx context.Context, stationID string) (string, error) {
	url := fmt.Sprintf("https://radiko.jp/v3/station/stream/pc_html5/%s.xml", stationID)
	status, xml, err := p.net.GetText(ctx, url, nil)
	if err != nil {
		return "", err
	}
	if status < 200 || status >= 300 {
		return "", fmt.Errorf("station stream xml failed: %d", status)
	}
	re := regexp.MustCompile(`<playlist_create_url>(.*?)</playlist_create_url>`)
	m := re.FindStringSubmatch(xml)
	if len(m) >= 2 {
		return strings.TrimSpace(m[1]), nil
	}
	// Older stations may omit this field; use the known default endpoint.
	return "https://tf-f-rpaa-radiko.smartstream.ne.jp/tf/playlist.m3u8", nil
}

// SegmentInput contains the server-side parameters required to request
// timeshift playlists and authorized chunklists.
type SegmentInput struct {
	StationID string
	FT        string
	TO        string
	Token     string
	AreaID    string
}

// BuildSegmentURLs iterates seek windows between FT and TO and collects media
// segment URLs from each chunklist response.
func (p *PlaylistBuilder) BuildSegmentURLs(ctx context.Context, in SegmentInput) ([]string, error) {
	const fixedSeek = 300
	base, err := p.playlistCreateURL(ctx, in.StationID)
	if err != nil {
		return nil, err
	}

	links := make([]string, 0, 1024)
	seek := in.FT
	endDate, err := util.ParseTimestamp(in.TO)
	if err != nil {
		return nil, err
	}

	for {
		seekDate, err := util.ParseTimestamp(seek)
		if err != nil {
			return nil, err
		}
		if !seekDate.Before(endDate) {
			break
		}

		playlistURL := fmt.Sprintf("%s?lsid=%s&station_id=%s&l=%d&start_at=%s&end_at=%s&type=b&ft=%s&to=%s&seek=%s",
			base, randomHex(16), in.StationID, fixedSeek, in.FT, in.TO, in.FT, in.TO, seek,
		)
		status, playlistText, err := p.net.GetText(ctx, playlistURL, map[string]string{
			"X-Radiko-AreaId":    in.AreaID,
			"X-Radiko-AuthToken": in.Token,
		})
		if err != nil {
			return nil, err
		}
		if status == 403 || strings.TrimSpace(playlistText) == "expired" || status < 200 || status >= 300 {
			// 403/expired/non-2xx all indicate the current seek window is unusable.
			return nil, fmt.Errorf("playlist request failed at seek=%s: %d", seek, status)
		}

		detailURL, err := firstDataLine(playlistText)
		if err != nil {
			return nil, err
		}
		chunkStatus, chunkText, err := p.net.GetText(ctx, detailURL, nil)
		if err != nil {
			return nil, err
		}
		if chunkStatus < 200 || chunkStatus >= 300 {
			return nil, fmt.Errorf("chunklist request failed: %d", chunkStatus)
		}
		links = append(links, allDataLines(chunkText)...)
		next, err := util.StepTimestamp(seek, fixedSeek)
		if err != nil {
			return nil, err
		}
		seek = next
	}

	return links, nil
}

func firstDataLine(m3u8 string) (string, error) {
	lines := allDataLines(m3u8)
	if len(lines) == 0 {
		return "", fmt.Errorf("m3u8 has no media lines")
	}
	return lines[0], nil
}

func allDataLines(m3u8 string) []string {
	out := make([]string, 0, strings.Count(m3u8, "\n")+1)
	start := 0
	for i := 0; i <= len(m3u8); i++ {
		if i < len(m3u8) && m3u8[i] != '\n' {
			continue
		}
		line := m3u8[start:i]
		start = i + 1
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		out = append(out, line)
	}
	return out
}
